### DISCUSSION TITLE
How can I get the distances between each subsequence？

### DISCUSSION THREAD

**Discussion by AllenYYs** [TDAmeritrade/stumpy#934] [Help: Coding & Implementations]
Basically, I know the matrix profile contains the distances between each subsequence of length m in X and its nearest neighbor subsequence in X.

Before the algorithm gets the final matrix profile, it must have calculated a distance matrix, which should contain the distance for each compared subsequence. I wonder if the Stumpy has saved these results, or I can just have access to the smallest Euclidean distance?

Thanks a lot for your help.

**Comment by JaKasb** [comment]
The MatrixProfile is the column-wise minimum-reduction of the distance matrix.

"How can I get the distances between each subsequence?"
You can re-compute the individual rows of the distance matrix with the function MASS(query, timeseries)
https://stumpy.readthedocs.io/en/latest/api.html#stumpy.mass
the variable "query" is your subsequence of interest with length m.

You can't access the full distance matrix inside the STAMP() algorithm, because the full distance matrix is too large to store in RAM.
The size of the distance matrix scales quadratically with the size of the input timeseries. 
e.g. if the input timeseries is 1 MegaByte the full distance matrix is 1 TeraByte...

**Comment by seanlaw** [comment]
@AllenYYs Welcome to the STUMPY community and thank you for your question!

> Before the algorithm gets the final matrix profile, it must have calculated a distance matrix, which should contain the distance for each compared subsequence. I wonder if the Stumpy has saved these results, or I can just have access to the smallest Euclidean distance?

So, while it may be true that ALL pairwise distances between each subsequence are computed at some point or another, only a list of the top-k (default is `k=1`) nearest neighbor is kept at any time. As @JaKasb pointed out, for any substantially large time series, you'll run out of memory (RAM) very quickly and this is the motivation behind the matrix profile. For a large majority of use cases, it is rarely useful to have/use the full `n x n` distance matrix and, instead, you should ask "what information am I really looking for that isn't (implicitly) captured by the one-nearest-neighbor and what information can be re-computed?"

Perhaps, you can describe what it is that you are trying to do (provide code and data where possible) and we can try to provide some further guidance/suggestions.

**Reply by seanlaw** [reply]
While this isn't supported by STUMPY (i.e., you are on your own if you use this), if your time series is reasonably short (i.e., `n < 1000`) AND you do not need to normalize your distances (i.e., `normalize=False`) then you might be able to do:

```
import numpy as np
from stumpy import core
from scipy.spatial.distance import cdist

T_A = np.random.rand(1000)
T_B = np.random.rand(1000)
m = 50

T_A[np.isinf(T_A)] = np.nan
T_B[np.isinf(T_A)] = np.nan

rolling_T_A = core.rolling_window(T_A, m)
rolling_T_B = core.rolling_window(T_B, m)

p = 2  # (non-normalized) Euclidean distance
distance_matrix = cdist(rolling_T_A, rolling_T_B, metric="minkowski", p=p)
```

**Reply by NimaSarajpoor** [reply]
If you want to normalize subseqs before computing the distance matrix and if the length of your time series is not long, you may use: "core.mass_distance_matrix". However, I think this is not supported as well. So, as @seanlaw mentioned, you can use it at your own risk.
